precompiles/erc20/approve.go

```diff
@@ -177,13 +178,18 @@ func (p Precompile) DecreaseAllowance(
 		err = sdkerrors.Wrapf(err, "allowance does not exist")
 	case subtractedValue != nil && subtractedValue.Cmp(allowance) < 0:
 		// case 3. subtractedValue positive and subtractedValue less than allowance -> update authorization
-		amount, err = p.decreaseAllowance(ctx, grantee, granter, subtractedValue, authorization)
+		amount, err = p.decreaseAllowance(ctx, grantee, granter, subtractedValue, authorization, expiration)
 	case subtractedValue != nil && subtractedValue.Cmp(allowance) == 0:
-		// case 4. subtractedValue positive and subtractedValue equal to allowance -> delete authorization
-		amount, err = p.decreaseAllowance(ctx, grantee, granter, subtractedValue, authorization)
+		// case 4. subtractedValue positive and subtractedValue equal to allowance -> remove spend limit for token and delete authorization if no other denoms are approved for
+		err = p.removeSpendLimitOrDeleteAuthorization(ctx, grantee, granter, authorization, expiration)
+		amount = common.Big0
 	case subtractedValue != nil && subtractedValue.Cmp(allowance) > 0:
-		// case 5. subtractedValue positive than higher than allowance -> return error
-		err = fmt.Errorf("subtracted value cannot be greater than existing allowance: %s > %s", subtractedValue, allowance)
+		// case 5. subtractedValue positive and subtractedValue higher than allowance -> return error
+		if allowance.Cmp(common.Big0) == 0 {
+			err = fmt.Errorf("allowance for token %s does not exist", p.tokenPair.Denom)
+		} else {
+			err = fmt.Errorf("subtracted value cannot be greater than existing allowance: %s > %s", subtractedValue, allowance)
+		}
```

Comment:
ditto we shouldn't have if/else inside the cases. If needed, please create new cases (5a. and 5b.) instead of reducing readability